File: Week 2/id_445/94.二叉树的中序遍历.cpp
/*
 * @lc app=leetcode.cn id=94 lang=cpp
 *
 * [94] 二叉树的中序遍历
 */
#include <vector>
#include <queue>
#include <stack>
using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
//solution 1 loop
//     vector<int> inorderTraversal(TreeNode* root) {
//         stack<TreeNode > rec;
//         vector<int> res;
//         TreeNode  * cur=root;

        
//         while (cur || !rec.empty())
//         {

//             while(cur!=NULL){
//             rec.push(*(cur));
//             cur=cur->left;
//             }
            
//                 if(!rec.empty()){
//                 res.push_back(rec.top().val);
//                 cur=rec.top().right;
//                 rec.pop(); 
//                 }
//         }
            
//         return res;

//     }
// };
// solution 2 recursion
 vector<int> inorderTraversal(TreeNode* root) {
        stack<TreeNode > rec;
        vector<int> res;
        TreeNode  * cur=root;

        recur(cur,res);
            
        return res;

    }
    void recur(TreeNode* node,vector<int> & res){
        if (node==NULL)
        {
            return ;
        }
        recur(node->left,res);
        res.push_back(node->val);
        recur(node->right,res);
    }
};

// @lc code=end

File: Week 2/id_445/94.二叉树的中序遍历.py
#
# @lc app=leetcode.cn id=94 lang=python3
#
# [94] 二叉树的中序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # solution 1 recursively
    # def inorderTraversal(self, root: TreeNode) -> List[int]:
    #     res=[]
    #     self.recur(root,res)
    #     return res
    # def recur(self,node,res):
    #     if node:
    #         self.recur(node.left,res)
    #         res.append(node.val)
    #         self.recur(node.right,res)
    # solution 2 iteratively
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res,stack=[],[]
        while True:
            while root:
                stack.append(root)
                root=root.left
            if not stack:
                return res
            node=stack.pop()
            res.append(node.val)
            root=node.right
        # return res

        
# @lc code=end

